contracts/PreToken.sol

```diff
@@ -0,0 +1,155 @@
+        preTokens are issued only to a closed group of contributors / team members.
+    If despite these something goes wrong then as a last resort a new pretoken contract can be recreated from agreements.
+
+    Some ERC20 functions are implemented so agreement owners can see their balances and use transfer in standard wallets.
+    Restrictions:
+      - only total balance can be transfered - effectively ERC20 transfer used to transfer agreement ownership
+      - only agreement holders can transfer
+        (i.e. can't transfer 0 amount if have no agreement to avoid polluting logs with Transfer events)
+      - transfer is only allowed to accounts without an agreement yet
+      - no approval and transferFrom ERC20 functions
+ */
+
+pragma solidity 0.4.24;
+import "./generic/SafeMath.sol";
+import "./generic/Restricted.sol";
+
+
+contract PreToken is Restricted {
+    using SafeMath for uint256;
+
+    uint public constant CHUNK_SIZE = 100;
+
+    string constant public name = "Augmint pretokens"; // solhint-disable-line const-name-snakecase
+    string constant public symbol = "APRE"; // solhint-disable-line const-name-snakecase
+    uint8 constant public decimals = 0; // solhint-disable-line const-name-snakecase
+
+    uint public totalSupply;
+
+    struct Agreement {
+        address owner;
+        uint balance;
+        uint32 discount; //  discountRate in parts per million , ie. 10,000 = 1%
+        uint32 valuationCap; // in USD (no decimals)
+    }
+
+    /* Agreement hash is the SHA-2 (SHA-256) hash of signed agreement document.
+         To generate:
+            OSX: shasum -a 256 agreement.pdf
+            Windows: certUtil -hashfile agreement.pdf SHA256 */
+    mapping(address => bytes32) public agreementOwners; // to lookup agrement by owner
+    mapping(bytes32 => Agreement) public agreements;
+
+    bytes32[] public allAgreements; // all agreements to able to iterate over
+
+    event Transfer(address indexed from, address indexed to, uint amount);
+
+    event NewAgreement(address owner, bytes32 agreementHash, uint32 discount, uint32 valuationCap);
+
+    function addAgreement(address owner, bytes32 agreementHash, uint32 discount, uint32 valuationCap)
+    external restrict("PreTokenSigner") {
+        require(owner != address(0), "owner must not be 0x0");
+        require(agreementOwners[owner] == 0x0, "owner must not have an aggrement yet");
+        require(agreementHash != 0x0, "agreementHash must not be 0x0");
+        require(discount > 0, "discount must be > 0");
+        require(agreements[agreementHash].discount == 0, "agreement must not exist yet");
+
+        agreements[agreementHash] = Agreement(owner, 0, discount, valuationCap);
+        agreementOwners[owner] = agreementHash;
+        allAgreements.push(agreementHash);
+
+        emit NewAgreement(owner, agreementHash, discount, valuationCap);
+    }
+
+    function issueTo(bytes32 agreementHash, uint amount) external restrict("PreTokenSigner") {
```

Comment:
Since `issueTo` and `burnFrom` are symmetrical, I would unify and use a signed int amount as parameter.

Reply:
I suggest to skip this. While it's valid point , I'm not sure it worth the effort: adding int into SafeMath (or check under/overflow here).  Different params for Transfer event depending on direction etc.  And for reviewing multisig scripts  it's more explicitly readable, i.e. issueTo is a common function while burnFrom shoudn't happen under normal circumstances